packages/ra-ui-materialui/src/form/SimpleFormIterator.js

```diff
@@ -64,7 +64,16 @@ export class SimpleFormIterator extends Component {
         // but redux-form doesn't provide one (cf https://github.com/erikras/redux-form/issues/2735)
         // so we keep an internal map between the field position and an autoincrement id
         this.nextId = 0;
-        this.ids = props.fields ? props.fields.map(() => this.nextId++) : [];
+
+        // We check whether we have a defaultValue (which must be an array) before checking
+        // the fields prop which will always be empty for a new record.
+        // Without it, our ids wouldn't match the default value and we would get key warnings
+        // on the CssTransition element inside our render method
+        this.ids = props.defaultValue
```

Comment:
I still don't understand why you don't check `props.fields` *before* checking `props.defaultValue`. Here is my rationale:

1. if there is a record, use the fields
2. if there is no record, use the default value

The 1st case doesn't work like that with your code.

By the way, since `this.nextId` is initialized just above, I think a faster way to do `props.fields.map(() => this.nextId++)` is `this.nextId = props.fields.length`.